system-addon/lib/TopStoriesFeed.jsm

```diff
@@ -79,11 +83,51 @@ this.TopStoriesFeed = class TopStoriesFeed {
       this.storiesLastUpdated = Date.now();
       // This is filtered so an update function can return true to retry on the next run
       this.contentUpdateQueue = this.contentUpdateQueue.filter(update => update());
+
+      this.setCache(STORIES_CACHE_KEY, this.stories);
     } catch (error) {
       Cu.reportError(`Failed to fetch content: ${error.message}`);
     }
   }
 
+  loadCachedStories() {
+    this.stories = this.getCachedStories();
+    if (this.stories && this.stories.length > 0) {
+      this.dispatchUpdateEvent(this.storiesLastUpdated, {rows: this.stories});
+      this.storiesLastUpdated = Date.now();
+    }
+  }
+
+  loadCachedTopics() {
+    const topics = this.getCachedTopics();
+    if (topics && topics.length > 0) {
+      this.dispatchUpdateEvent(this.topicsLastUpdated, {topics, read_more_endpoint: this.read_more_endpoint});
+      this.topicsLastUpdated = Date.now();
+    }
+  }
+
+  getCachedStories() {
+    return this.getCached(STORIES_CACHE_KEY);
+  }
+
+  getCachedTopics() {
+    return this.getCached(TOPICS_CACHE_KEY);
+  }
+
+  getCached(key) {
+    let results = [];
+    try {
+      results = JSON.parse(new Prefs().get(key));
```

Comment:
We really should move away from this anti-pattern of `new Prefs` see #3431. You should be able to do `this.store.getState().Prefs.values[key]` from here.

Reply:
yeah, did this because I was afraid of a race condition with the PrefsFeed. Or is PrefsFeed guaranteed to be first?

Reply:
Nod. I noted in #3431 that `PrefsFeed` currently `INIT`s after some feeds, but luckily for us here, `T`opStories comes after `P`refsFeed. ;)